### PR TITLE
Make out-of-tree builds easier

### PR DESCRIPTION
**Pull request by bradjc**
I'm trying to use Tock as a dependency for the Tock bootloader. Here is an example Cargo.toml file:

```
[package]
name = "nano33-bootloader"
version = "0.1.0"
authors = ["Tock Project Developers <tock-dev@googlegroups.com>"]
build = "build.rs"
edition = "2018"

[dependencies]
cortexm4 = { git = "https://github.com/tock/tock", branch = "master" }
capsules = { git = "https://github.com/tock/tock", branch = "master" }
kernel = { git = "https://github.com/tock/tock", branch = "master" }
nrf52 = { git = "https://github.com/tock/tock", branch = "master" }
nrf52840 = { git = "https://github.com/tock/tock", branch = "master" }
components = { git = "https://github.com/tock/tock", branch = "master" }
nrf52_components = { git = "https://github.com/tock/tock", branch = "master" }

bootloader = { path = "../../bootloader" }
bootloader_nrf52 = { path = "../../chips/bootloader_nrf52" }
bootloader_cortexm = { path = "../../arch/bootloader_cortexm" }
```

This however causes cargo to fetch all Tock submodules and all of _their_ submodules. This takes a very long time (10-15 minutes on my computer). Worse, `cmocka.git` doesn't work for some reason, so I get the following output:

```shell
$ make
    Updating git repository `https://github.com/tock/tock`
    Updating git submodule `https://git.qemu.org/git/qemu.git`
    Updating git submodule `https://git.qemu.org/git/capstone.git`
    Updating git submodule `https://git.qemu.org/git/dtc.git`
    Updating git submodule `https://git.qemu.org/git/meson.git`
    Updating git submodule `https://git.qemu.org/git/QemuMacDrivers.git`
    Updating git submodule `https://git.qemu.org/git/SLOF.git`
    Updating git submodule `https://git.qemu.org/git/edk2.git`
    Updating git submodule `https://github.com/google/brotli`
    Updating git submodule `https://github.com/hillbig/esaxx`
    Updating git submodule `https://github.com/y-256/libdivsufsort.git`
    Updating git submodule `https://github.com/openssl/openssl`
    Updating git submodule `https://boringssl.googlesource.com/boringssl`
    Updating git submodule `https://github.com/krb5/krb5`
    Updating git submodule `https://github.com/pyca/cryptography.git`
    Updating git submodule `https://github.com/google/brotli`
    Updating git submodule `https://github.com/hillbig/esaxx`
    Updating git submodule `https://github.com/y-256/libdivsufsort.git`
    Updating git submodule `https://github.com/kkos/oniguruma`
    Updating git submodule `https://github.com/ucb-bar/berkeley-softfloat-3.git`
    Updating git submodule `https://git.cryptomilk.org/projects/cmocka.git`
warning: spurious network error (2 tries remaining): early EOF; class=Net (12); code=Eof (-20)
warning: spurious network error (1 tries remaining): early EOF; class=Net (12); code=Eof (-20)
error: failed to get `capsules` as a dependency of package `nano33-bootloader v0.1.0 (/Users/bradjc/git/tock-bootloader/boards/nano33-bootloader)`

Caused by:
  failed to load source for dependency `capsules`

Caused by:
  Unable to update https://github.com/tock/tock?branch=master

Caused by:
  failed to update submodule `tools/qemu`

Caused by:
  failed to update submodule `roms/edk2`

Caused by:
  failed to update submodule `UnitTestFrameworkPkg/Library/CmockaLib/cmocka`

Caused by:
  failed to fetch submodule `UnitTestFrameworkPkg/Library/CmockaLib/cmocka` from https://git.cryptomilk.org/projects/cmocka.git

Caused by:
  network failure seems to have happened
  if a proxy or similar is necessary `net.git-fetch-with-cli` may help here
  https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli

Caused by:
  early EOF; class=Net (12); code=Eof (-20)
make: *** [/Users/bradjc/git/tock-bootloader/target/thumbv7em-none-eabi/release/nano33-bootloader] Error 101
```

I have no idea why cargo cannot clone cmocka. I've been able to work around this by cloning cmocka elsewhere and copying it to where cargo is trying to clone it. Then cargo continues after it.

I don't think that I'm doing anything unusual to cause this, but this is a somewhat big ask for anyone to do when trying to use Tock from github as a dependency. I also don't think we intended to make out-of-tree development difficult when adding the submodule.

### Testing Strategy

Todo? I'm not sure how to test this without merging it first.

### TODO or Help Wanted

I have no idea if this is the best way to fix this. However, I've now been stung by this twice, so I wanted to make it sure it got written down somewhere.

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by lschuermann** [comment]
I ran into this exact same issue multiple times as well. I'm pretty sure that this should be considered a tooling issue in Cargo, and submodules are not the actual issue here. Removing the submodule should of course fix the issue nonetheless.

I feel like I originally misunderstood the reason why we need to have QEMU in Tock anyways. Given that I can only find references to it in the qemu-runner, and it seems like we're just building it for CI, maybe it's safe to assume that for all relevant CI jobs the user is required to have an up-to-date & patched version of QEMU installed on their system? For the actual CI we might be fine just pinning the QEMU Git revision in the GitHub actions workflow and executing the build steps there?

Given this was an option in the original discussion around submodules, I'd vouch to not have any type of script or Makefile try to download and build QEMU automatically on a developer-machine, or only after an explicit prompt/warning for the reasons outlined by [@gendx here](https://github.com/tock/tock/issues/1865#issuecomment-630976331). Besides that, the fact that those build steps even work on a random developer machine would be pure luck and I'd prefer to build a patched QEMU using my package manager.

**Comment by hudson-ayers** [comment]
I think we should keep the QEMU CI tests but am all for replacing the submodule with just a Makefile rule to clone some pinned QEMU commit.

I know the original dissent to that was "its basically just manually recreating the functionality of a submodule", but after some experience with the submodule I disagree, for the following reasons:

1. When a git repository is used as a cargo dependency, cargo automatically recursively initializes all submodules. This means that any out-of-tree users of Tock are forced into installing QEMU and all of its submodules, which takes a long time and a bunch of space on disk. There is no option to just not initialize the submodule. This also means that if any of the submodules in the QEMU repository fail to install (as happened to Brad here), that it becomes impossible to use Tock as a git dependency through Cargo.

2. If QEMU is not a submodule but instead just installed via a `Makefile` rule, then the `tools/qemu` directory can be included in `.gitignore`. This cannot be done if it is a submodule, which means that tools like `ripgrep` which respect your `.gitignore` now search the entire QEMU codebase rather than just the Tock codebase. There are ways around this by using `ripgrep` specific `.ignore` files, but it is much less ergonomic for new users. And I think that tools like `ripgrep` are quite popular among rust developers.

3. I have seen occassions where users accidentally update the qemu submodule back to an older version as part of a PR, which then has to be manually reverted. This would be much less likely to happen if not for the confusing UX of submodules

**Comment by lschuermann** [comment]
> I think we should keep the QEMU CI tests but am all for replacing the submodule with just a Makefile rule to clone some pinned QEMU commit.

What about performing the actual QEMU build steps in a GitHub Actions build step (with pinned Git revision)? I don't feel like we should really be supporting building QEMU as part of the Tock Makefile infrastructure at all, as that comes with a lot of edge-cases and some people aren't comfortable with that. I think it's a fair requirement to state that a recent (or even patched) version is required to run the QEMU-CI tests locally. _If_ this should be included in the Makefile, we should at least detect whether we're running on GitHub Actions and otherwise prompt the user whether they'd like building QEMU to be attempted automatically ([for the security concerns mentioned by @gendx](https://github.com/tock/tock/issues/1865#issuecomment-630976331)).

**Comment by hudson-ayers** [comment]
> What about performing the actual QEMU build steps in a GitHub Actions build step (with pinned Git revision)? I don't feel like we should really be supporting building QEMU as part of the Tock Makefile infrastructure at all, as that comes with a lot of edge-cases and some people aren't comfortable with that. I think it's a fair requirement to state that a recent (or even patched) version is required to run the QEMU-CI tests locally. _If_ this should be included in the Makefile, we should at least detect whether we're running on GitHub Actions and otherwise prompt the user whether they'd like building QEMU to be attempted automatically ([for the security concerns mentioned by @gendx](https://github.com/tock/tock/issues/1865#issuecomment-630976331)).

Running `make ci-setup-qemu` already has exactly that behavior -- it prompts unless it detects it is running on GitHub Actions -- I am just proposing we replace the contents of that Make rule with cloning the repo at a certain commit hash rather than calling `git submodule sync; git submodule update --init`.

This mirrors how we have make rules to install `markdown-toc` or `libusb` for people who want to run those tests or those portions of CI.

**Comment by ppannuto** [comment]
I find these arguments compelling, particularly the argument of cargo automatically recursively cloning submodules being unnecessarily burdensome for out of tree users.

@hudson-ayers, do you have the other half of this (the clone as a make rule) in a branch or anything somewhere?

**Comment by hudson-ayers** [comment]
I do, but it does not work completely yet. 

I noticed some issues with the existing qemu rule - even on github actions, the following 2 lines are printed at the beginning of every run:
```
bash:  status=$(git submodule status -- tools/qemu); [[ "${status:0:1}" != "" ]] && cd tools/qemu/build && make -q riscv32-softmmu && echo yes: No such file or directory
bash:  [[ $(cksum tools/qemu-runner/opentitan-boot-rom.elf | cut -d" " -f1) == "328682170" ]] && echo yes: No such file or directory
```
which suggests something is not working as expected.

I was planning to come back to this after the HotOS deadline (February 3)

**Comment by bradjc** [comment]
From tock-bootloader PR (https://github.com/tock/tock-bootloader/pull/21#issuecomment-770198391):

> It might be because I had to use my local version (the comments in the "Development" part of Cargo.toml) as pulling the tock repository from github clones it recursively and fails due to some qemu repository dependency. I thought is was my computer's issue, but the running the build in github workflows yields the same error https://github.com/WyliodrinEmbeddedIoT/tock-bootloader/runs/1750903959?check_suite_focus=true

**Comment by hudson-ayers** [comment]
I pushed what I believe to be an appropriate fix for the QEMU CI that no longer relies on the presence of a submodule. It also includes a fix for `ci-setup-helper`, which printed errors as part of operation on ubuntu (as can be seen in the ouput of any of the old GA workflows).

This still needs to be updated to use whatever the correct current QEMU commit is

File: Makefile [hunk 1]
```diff
@@ -494,12 +495,13 @@ ci-job-miri: ci-setup-miri
...
 ### ci-runner-github-qemu jobs:
 
+QEMU_COMMIT_HASH=a05f8ecd88f15273d033b6f044b850a8af84a5b8
 define ci_setup_qemu_riscv
 	$(call banner,CI-Setup: Build QEMU)
 	@# Use the latest QEMU as it has OpenTitan support
 	@printf "Building QEMU, this could take a few minutes\n\n"
-	@git submodule sync; git submodule update --init
-	@cd tools/qemu; ../qemu/configure --target-list=riscv32-softmmu --disable-linux-io-uring --disable-libdaxctl;
+  @git clone https://github.com/qemu/qemu ./tools/qemu 2>/dev/null || echo "qemu already cloned, checking out"
```
Comment (ppannuto) [inline Makefile:503]: Since we're minimizing, a slick thing to do might be to just pull down what's needed: https://stackoverflow.com/questions/31278902/how-to-shallow-clone-a-specific-commit-with-depth-1

**Review by ppannuto — APPROVED** [review]
This looks good to me!

File: Makefile (hunk 1 above)
Comment (alistair23) [reply Makefile:503]: The indentation looks wrong here.

Comment (hudson-ayers) [reply Makefile:503]: Ah, it turns out this entire file uses tabs instead of spaces :/

Comment (jrvanwhy) [reply Makefile:503]: > Ah, it turns out this entire file uses tabs instead of spaces :/

`make` requires the indentation to be tabs.

Comment (hudson-ayers) [reply Makefile:503]: @ppannuto I tried the approach in the answer and did not get it to work. I may have just done something wrong. But the comments under the accepted answer suggest that GitHub repositories do not have the correct features enabled to support shallow cloning from a specific commit. Other comments suggest that this approach also creates a dependence on very new versions of git (e.g. sometime in 2019) which we might prefer to avoid anyways.

Feel free to add a commit that does this if you can get it to work though

**Review by alistair23 — APPROVED** [review]

File: Makefile (hunk 1 above)
Comment (ppannuto) [reply Makefile]: That sounds like it might be more fragile than it is worth for this use case

**Review by hudson-ayers — APPROVED** [review]

**Comment by carnotweat** [comment]
> I'm trying to use Tock as a dependency for the Tock bootloader. Here is an example Cargo.toml file:
> 
> ```
> [package]
> name = "nano33-bootloader"
> version = "0.1.0"
> authors = ["Tock Project Developers <tock-dev@googlegroups.com>"]
> build = "build.rs"
> edition = "2018"
> 
> [dependencies]
> cortexm4 = { git = "https://github.com/tock/tock", branch = "master" }
> capsules = { git = "https://github.com/tock/tock", branch = "master" }
> kernel = { git = "https://github.com/tock/tock", branch = "master" }
> nrf52 = { git = "https://github.com/tock/tock", branch = "master" }
> nrf52840 = { git = "https://github.com/tock/tock", branch = "master" }
> components = { git = "https://github.com/tock/tock", branch = "master" }
> nrf52_components = { git = "https://github.com/tock/tock", branch = "master" }
> 
> bootloader = { path = "../../bootloader" }
> bootloader_nrf52 = { path = "../../chips/bootloader_nrf52" }
> bootloader_cortexm = { path = "../../arch/bootloader_cortexm" }
> ```
> 
> This however causes cargo to fetch all Tock submodules and all of _their_ submodules. This takes a very long time (10-15 minutes on my computer). Worse, `cmocka.git` doesn't work for some reason, so I get the following output:
> 
> ```shell
> $ make
>     Updating git repository `https://github.com/tock/tock`
>     Updating git submodule `https://git.qemu.org/git/qemu.git`
>     Updating git submodule `https://git.qemu.org/git/capstone.git`
>     Updating git submodule `https://git.qemu.org/git/dtc.git`
>     Updating git submodule `https://git.qemu.org/git/meson.git`
>     Updating git submodule `https://git.qemu.org/git/QemuMacDrivers.git`
>     Updating git submodule `https://git.qemu.org/git/SLOF.git`
>     Updating git submodule `https://git.qemu.org/git/edk2.git`
>     Updating git submodule `https://github.com/google/brotli`
>     Updating git submodule `https://github.com/hillbig/esaxx`
>     Updating git submodule `https://github.com/y-256/libdivsufsort.git`
>     Updating git submodule `https://github.com/openssl/openssl`
>     Updating git submodule `https://boringssl.googlesource.com/boringssl`
>     Updating git submodule `https://github.com/krb5/krb5`
>     Updating git submodule `https://github.com/pyca/cryptography.git`
>     Updating git submodule `https://github.com/google/brotli`
>     Updating git submodule `https://github.com/hillbig/esaxx`
>     Updating git submodule `https://github.com/y-256/libdivsufsort.git`
>     Updating git submodule `https://github.com/kkos/oniguruma`
>     Updating git submodule `https://github.com/ucb-bar/berkeley-softfloat-3.git`
>     Updating git submodule `https://git.cryptomilk.org/projects/cmocka.git`
> warning: spurious network error (2 tries remaining): early EOF; class=Net (12); code=Eof (-20)
> warning: spurious network error (1 tries remaining): early EOF; class=Net (12); code=Eof (-20)
> error: failed to get `capsules` as a dependency of package `nano33-bootloader v0.1.0 (/Users/bradjc/git/tock-bootloader/boards/nano33-bootloader)`
> 
> Caused by:
>   failed to load source for dependency `capsules`
> 
> Caused by:
>   Unable to update https://github.com/tock/tock?branch=master
> 
> Caused by:
>   failed to update submodule `tools/qemu`
> 
> Caused by:
>   failed to update submodule `roms/edk2`
> 
> Caused by:
>   failed to update submodule `UnitTestFrameworkPkg/Library/CmockaLib/cmocka`
> 
> Caused by:
>   failed to fetch submodule `UnitTestFrameworkPkg/Library/CmockaLib/cmocka` from https://git.cryptomilk.org/projects/cmocka.git
> 
> Caused by:
>   network failure seems to have happened
>   if a proxy or similar is necessary `net.git-fetch-with-cli` may help here
>   https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli
> 
> Caused by:
>   early EOF; class=Net (12); code=Eof (-20)
> make: *** [/Users/bradjc/git/tock-bootloader/target/thumbv7em-none-eabi/release/nano33-bootloader] Error 101
> ```
> 
> I have no idea why cargo cannot clone cmocka. I've been able to work around this by cloning cmocka elsewhere and copying it to where cargo is trying to clone it. Then cargo continues after it.
> 
> I don't think that I'm doing anything unusual to cause this, but this is a somewhat big ask for anyone to do when trying to use Tock from github as a dependency. I also don't think we intended to make out-of-tree development difficult when adding the submodule.
> 
> ### Testing Strategy
> Todo? I'm not sure how to test this without merging it first.
> 
> ### TODO or Help Wanted
> I have no idea if this is the best way to fix this. However, I've now been stung by this twice, so I wanted to make it sure it got written down somewhere.
> 
> ### Documentation Updated
> * [ ]  Updated the relevant files in `/docs`, or no updates are required.
> 
> ### Formatting
> * [x]  Ran `make prepush`.

I am also experiencing a spurious network error 
How to reproduce ?
https://git.sr.ht/~carnotweat/modshelltovm/tree/main/item/flake.nix#L98
```
nix build #default -j 50 --impure   --show-trace     
error: builder for '/nix/store/m6zx4rl9anpk0zji61slz1a8ciqybzky-cargo-ws.drv' failed with exit code 101;
       last 25 log lines:
       > The make system also drives all continuous integration and testing:
       > /nix/store/vpvy79k1qq02p1vyqjk6nb89gwhxqvyb-bash-5.2p32/bin/bash: line 1: tput: command not found
       > /nix/store/vpvy79k1qq02p1vyqjk6nb89gwhxqvyb-bash-5.2p32/bin/bash: line 1: tput: command not found
       >          prepush: Fast checks to run before pushing changes upstream
       >           ci-all: Run all continuous integration tests (possibly slow!)
       >          ci-help: More information on Tock CI and testing
       >
       > /nix/store/vpvy79k1qq02p1vyqjk6nb89gwhxqvyb-bash-5.2p32/bin/bash: line 1: tput: command not found
       > /nix/store/vpvy79k1qq02p1vyqjk6nb89gwhxqvyb-bash-5.2p32/bin/bash: line 1: tput: command not found
       > Happy Hacking!
       > Running phase: installPhase
       >     Updating crates.io index
       > warning: spurious network error (3 tries remaining): [6] Could not resolve hostname (Could not resolve host: index.crates.io)
       > warning: spurious network error (2 tries remaining): [6] Could not resolve hostname (Could not resolve host: index.crates.io)
       > warning: spurious network error (1 tries remaining): [6] Could not resolve hostname (Could not resolve host: index.crates.io)
       > error: failed to get `ghash` as a dependency of package `capsules-aes-gcm v0.1.0 (/build/source/capsules/aes_gcm)`
       >
       > Caused by:
       >   download of config.json failed
       >
       > Caused by:
       >   failed to download from `https://index.crates.io/config.json`
       >
       > Caused by:
       >   [6] Could not resolve hostname (Could not resolve host: index.crates.io)
       For full logs, run 'nix log /nix/store/m6zx4rl9anpk0zji61slz1a8ciqybzky-cargo-ws.drv'.
 
```
to reproduce - 
```
nix build #default -j 50 --impure   --show-trace
```
in my repository
